metis-repository/src/main/java/eu/europeana/metis/repository/rest/OaiPmhController.java

```diff
@@ -0,0 +1,140 @@
+import org.springframework.http.HttpStatus;
+import org.springframework.http.MediaType;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.ResponseBody;
+import org.springframework.web.bind.annotation.ResponseStatus;
+import org.springframework.web.bind.annotation.RestController;
+import org.springframework.web.server.ResponseStatusException;
+
+/**
+ * Controller for OAI-PMH harvesting.
+ */
+@RestController
+@Tags(@Tag(name = OaiPmhController.CONTROLLER_TAG_NAME,
+    description = "Controller providing access to OAI-PMH harvesting functionality."))
+@Api(tags = OaiPmhController.CONTROLLER_TAG_NAME)
+public class OaiPmhController {
+
+  public static final String CONTROLLER_TAG_NAME = "OaiPmhController";
+  private static final Logger LOGGER = LoggerFactory.getLogger(OaiPmhController.class);
+
+  private RecordDao recordDao;
+
+  @Autowired
+  void setRecordDao(RecordDao recordDao) {
+    this.recordDao = recordDao;
+  }
+
+  @GetMapping(value = RestEndpoints.OAI_ENDPOINT, produces = {MediaType.APPLICATION_XML_VALUE})
+  @ResponseStatus(HttpStatus.OK)
+  @ResponseBody
+  @ApiOperation(value = "OAI endpoint (supporting only the ListIdentifiers and GetRecord verbs)")
+  @ApiResponses(value = {@ApiResponse(code = 400, message = "Illegal OAI request"),
+          @ApiResponse(code = 404, message = "Unknown dataset or record ID"),
+          @ApiResponse(code = 500, message = "Error processing the request")})
+  public String oaiPmh(
+          @ApiParam(value = "The verb (ListIdentifiers or GetRecords)", required = true) @QueryParam("verb") String verb,
+          @ApiParam(value = "The set (required for ListIdentifiers)") @QueryParam("set") String set,
+          @ApiParam(value = "The metadataPrefix (only 'edm' is supported.)", required = true) @QueryParam("metadataPrefix") String metadataPrefix,
+          @ApiParam(value = "The record identifier (required for GetRecord)") @QueryParam("identifier") String identifier) {
+
+    // Check the metadata prefix
+    if (!"edm".equals(metadataPrefix)) {
+      throw new ResponseStatusException(HttpStatus.BAD_REQUEST,
+              "Unsupported metadataPrefix value: " + metadataPrefix);
+    }
+
+    // Check the verb and delegate
+    final Verb verbResult;
+    if ("ListIdentifiers".equals(verb)) {
+      verbResult = listIdentifiers(set);
+    } else if ("GetRecord".equals(verb)) {
+      verbResult = getRecord(identifier);
+    } else {
+      throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "Unsupported verb: " + verb);
+    }
+
+    // Compile the result
+    final OAIPMH result = new OAIPMH().withVerb(verbResult).withResponseDate(new Date())
+            .withRequest(new Request(RestEndpoints.OAI_ENDPOINT).withVerbType(verbResult.getType()));
+    try {
+      ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+      final WriterContext context = new WriterContext(Granularity.Day, new SimpleResumptionTokenFormat());
```

Comment:
I'm not familiar with `WriterContext`, what's its purpose in here?

Reply:
It's an OAI specific term. I'm really only using it to specify that we want to send the datestamps as dates only (and not include a time component). I think that will be more beneficial to our testing.